Multiple overlay support

beniwtv opened the pull request:
This is multiple OpenVR overlays support!

Not much changed, but there are two breaking changes:
- create_overlay() returns an ID for the created overlay now, which
  has to be used in subsequent overlay function calls
- create_overlay() now also needs the texture ID of the Viewport to be used as texture, which you can get it in Godot through the VisualServer.

Example:
```
# Wait for everything to be ready - in case Viewport hasn't been drawn yet
yield(get_tree(), "idle_frame")
yield(get_tree(), "idle_frame")
yield(get_tree(), "idle_frame")
yield(get_tree(), "idle_frame")

# Get texture ID
var texid_overlay1 = VisualServer.texture_get_texid(VisualServer.viewport_get_texture($VRViewport1.get_viewport_rid()))
var texid_overlay2 = VisualServer.texture_get_texid(VisualServer.viewport_get_texture($VRViewport2.get_viewport_rid()))
	
OpenVROverlay = preload("res://addons/godot-openvr/OpenVROverlay.gdns").new()
var overlay_id1 = OpenVROverlay.create_overlay("identifier", "Friendly name", texid_overlay1) 
var overlay_id2 = OpenVROverlay.create_overlay("identifier2", "Friendly name2", texid_overlay2) 
```

:)

BastiaanOlij commented:
I think we need a nicer way to do this.. Instead of providing the texture id to the module we can provide a NodePath to the viewport.

Then inside the module we can simply check if the viewport related to the overlay has a texture id. We can implement a lazy loading approach here where when we're accessing the overlay, we check if the texture id has none zero, if it has changed since last time, and whether we already have created the overlay, and either return that overlay or create a new overlay as a result.

It does mean that the code that uses the overlay should check if there is a valid overlay.

beniwtv commented:
> I think we need a nicer way to do this.. Instead of providing the texture id to the module we can provide a NodePath to the viewport

Yes I think that could work. I'd have to query the VisualServer inside the module instead and check if it matches the texid given to godot_arvr_commit_for_eye(), and only then set the texture.

> Then inside the module we can simply check if the viewport related to the overlay has a texture id. We can implement a lazy loading approach here where when we're accessing the overlay, we check if the texture id has none zero, if it has changed since last time, and whether we already have created the overlay, and either return that overlay or create a new overlay as a result.
> 
> It does mean that the code that uses the overlay should check if there is a valid overlay.

OpenVR already does not display the overlay until a valid texture is set - so I think this would not be needed?

beniwtv commented:
I have changed create_overlay() to require a path to the Viewport now. This eliminates the need for weird idle_frame yields we had before, and is easier to understand.

I left it as a new commit for now so you can easier check what I changed, will squash them before merge.

One small nitpick remains: We now have to add_child() the OpenVROverlay class in GDscript that using this class, else the module doesn't seem to have access to the scene tree to resolve the Viewport.

Do you know any way around that?

BastiaanOlij commented:
Already discussed this on discord but just to record it on the PR. The idea we discussed to maybe take this one step further is to have the Overlay class inherit from viewport instead of being a subclass of node.

This would allow us to apply the script to a viewport node and turn that viewport into an overlay. This changes the object so we will have an instance for every overlay but it also means the interface becomes easier to understand.

beniwtv commented:
Working on this as we speak :)

beniwtv commented:
So I think this is now easier than ever! \o/

Basically, you just have to drag OpenVROverlay.gdns into the script property of a Viewport and it will create a basic overlay already visible in SteamVR via the following steps:
- Create an overlay with your Godot game/apps's name + number (derived from the number of overlays present in the game/app) as name for the overlay in SteamVR
- Set the viewport to arvr = true
- Apply a base transform to the overlay (static) to make it appear in the VR world, with the Viewport's contents on it
- Make the overlay visible (or not, depends on the overlay_visible property but defaults to true) and set the width in meters (depends on the overlay_width_in_meters property, default 1 meter)

It still allows, of course, to set it's visibility, width and transform after the fact via properties/functions.
